docs/sdg/sdg-api-interface.md

```diff
@@ -0,0 +1,71 @@
+# SDG Library Design
+
+
+## Objective
+Library called `instructlab-sdg` that can be called per seed example that includes question and answer pairs, and context for grounded skills.
+
+## Structure of the SDG Library
+
+We propose the following structure for the SDG library. There will be config files that contain all the prompt templates for the pipelines. 
+
+
+```
+- src/instructlab/sdg/
+    - configs/ 
+        - gen_q.yaml 
+        - gen_a.yaml
+        - ...
+    - init.py 
+    - block.py 
+    - llmblock.py 
+    - pipeline.py 
+    - sdg.py
+```
+
+
+## CLI
+
+The CLI client uses the instructlab SDG library and provides it a run configuration with input parameters.
+
+
+```
+# cli_driver.py
+
+import sdg
+
+client = openai_client(endpoint)
+cli_sdg = sdg(run_config, client)  # run config has all the variables like num_samples, pipelinesteps etc
+generated_samples = cli_sdg.generate()
+```
+
+The run configuration contains the parameters needed to run the SDG code library as well as points to the templates needed to run the SDG code as well as the prompt template and the default model system prompt template.
+
+* num_samples is the number of synthetic samples that you wish to generate per seed example
+* max_retry is the maximum number of non-greedy retries you want to make if the num_samples is not reached. The number of samples in the generated output will be the samples achieved until max_retry is reached
+* Pipeline steps contains the steps that you want to invoke in the SDG pipeline and the prompt configurations per step. The variable names of the blocks can be anything and the prompt configs must be compatible with the teacher model
+*  max_new_tokens is the maximum number of tokens we want to generate. In other words, the size of the output sequence, not including the tokens in the prompt. 
+*  model_name is the teacher model we would want to use to generate the synthetic data.
+*  model_template and stop_token are the paramaters for generation template.
+
+```
+# run_config.yaml
+
+num_samples : 30
+max_retry : 5
+pipeline_steps:
+  gen_q:
+    prompt_template: "configs/gen_q.yaml"
+  filter_q:
+   prompt_template: "configs/filter_q.yaml" 
+max_new_tokens: 10000 
+# model parameters for generation
+model_name: mistralai/Mixtral-8x7B-Instruct-v0.1
```

Comment:
I think there's a need to configure the model name on a per-stage basis. We may have an endpoint configured with more than one model and we want to hit different ones depending on the stage.

Reply:
This was marked as resolved, but I don't think it got moved.

Reply:
@russellb made updates to this to accept a client, model-name and default model prompt

Reply:
@oindrillac but it's still not specified on a per-stage basis?

Reply:
the value will be a class variable that gets passed on when an object is initialized. we have also decided to keep the run config in a .py file as per your suggestion let me update the doc

Reply:
PTAL @russellb

Reply:
cc: @shivchander @aakankshaduggal